api/turing/batch/runner/ensembling.go

```diff
@@ -0,0 +1,309 @@
+package batchrunner
+
+import (
+	"time"
+
+	mlp "github.com/gojek/mlp/client"
+	batchcontroller "github.com/gojek/turing/api/turing/batch/controller"
+	"github.com/gojek/turing/api/turing/imagebuilder"
+	"github.com/gojek/turing/api/turing/internal/testutils"
+	"github.com/gojek/turing/api/turing/labeller"
+	"github.com/gojek/turing/api/turing/log"
+	"github.com/gojek/turing/api/turing/models"
+	"github.com/gojek/turing/api/turing/service"
+)
+
+type ensemblingJobRunner struct {
+	ensemblingController      batchcontroller.EnsemblingController
+	ensemblingJobService      service.EnsemblingJobService
+	mlpService                service.MLPService
+	imageBuilder              imagebuilder.ImageBuilder
+	environment               string
+	batchSize                 int
+	maxRetryCount             int
+	imageBuildTimeoutDuration time.Duration
+}
+
+// NewBatchEnsemblingJobRunner creates a new batch ensembling job runner
+// This service controls the orchestration of batch ensembling jobs.
+func NewBatchEnsemblingJobRunner(
+	ensemblingController batchcontroller.EnsemblingController,
+	ensemblingJobService service.EnsemblingJobService,
+	mlpService service.MLPService,
+	imageBuilder imagebuilder.ImageBuilder,
+	environment string,
+	batchSize int,
+	maxRetryCount int,
+	imageBuildTimeoutDuration time.Duration,
+) BatchJobRunner {
+	return &ensemblingJobRunner{
+		ensemblingController:      ensemblingController,
+		ensemblingJobService:      ensemblingJobService,
+		mlpService:                mlpService,
+		imageBuilder:              imageBuilder,
+		environment:               environment,
+		batchSize:                 batchSize,
+		maxRetryCount:             maxRetryCount,
+		imageBuildTimeoutDuration: imageBuildTimeoutDuration,
+	}
+}
+
+func (r *ensemblingJobRunner) Run() {
+	r.processJobs()
+	r.updateStatus()
+}
+
+func (r *ensemblingJobRunner) updateStatus() {
+	// Here we want to check all non terminal cases.
+	// i.e. JobBuildingImage but with updated_at timeout
+	//		JobRunning
+	optionsJobRunning := service.EnsemblingJobListOptions{
+		PaginationOptions: service.PaginationOptions{
+			Page:     testutils.NullableInt(1),
+			PageSize: &r.batchSize,
+		},
+		Statuses: []models.Status{
+			models.JobRunning,
+		},
+	}
+	r.processEnsemblingJobs(optionsJobRunning)
+
+	// To ensure that we only pickup jobs that weren't picked up, we check the last updated at.
+	updatedAtBefore := time.Now().Add(r.imageBuildTimeoutDuration * -1)
+	optionsJobBuildingImage := service.EnsemblingJobListOptions{
+		PaginationOptions: service.PaginationOptions{
+			Page:     testutils.NullableInt(1),
+			PageSize: &r.batchSize,
+		},
+		Statuses: []models.Status{
+			models.JobBuildingImage,
+		},
+		UpdatedAtBefore:    &updatedAtBefore,
+		RetryCountLessThan: &r.maxRetryCount,
+	}
+	r.processEnsemblingJobs(optionsJobBuildingImage)
+}
+
+func (r *ensemblingJobRunner) processEnsemblingJobs(queryOptions service.EnsemblingJobListOptions) {
```

Comment:
Actually this is rather awkward because some options like the updatedAtBefore wont be used in the JobRunning case (where we are checking for the status of the spark application as opposed to the image building section).

I split it into different options because both require different combinations, i.e the case of JobRunning and the JobBuilding case. I think it would make sense if we had overloaded functions/methods in Go but it does not seem like that is the case.